File: firefly-example/src/main/java/com/firefly/example/reactive/coffee/store/dao/impl/UserDAOImpl.java
package com.firefly.example.reactive.coffee.store.dao.impl;

import com.firefly.$;
import com.firefly.annotation.Component;
import com.firefly.annotation.Inject;
import com.firefly.example.reactive.coffee.store.dao.UserDAO;
import com.firefly.example.reactive.coffee.store.model.User;
import com.firefly.reactive.adapter.db.ReactiveTransactionalManager;
import reactor.core.publisher.Mono;

/**
 * @author Pengtao Qiu
 */
@Component
public class UserDAOImpl implements UserDAO {

    @Inject
    private ReactiveTransactionalManager db;

    @Override
    public Mono<User> getByName(String name) {
        if (!$.string.hasText(name)) {
            return Mono.error(new IllegalArgumentException("The username is required"));
        }

        String sql = "select * from `coffee_store`.`user` where `name` = ?";
        return db.execSQL(c -> c.queryForObject(sql, User.class, name));
    }
}

File: firefly-example/src/main/kotlin/com/firefly/example/kotlin/coffee/store/dao/impl/UserDAOImpl.kt
package com.firefly.example.kotlin.coffee.store.dao.impl

import com.firefly.annotation.Component
import com.firefly.annotation.Inject
import com.firefly.example.kotlin.coffee.store.dao.UserDAO
import com.firefly.example.kotlin.coffee.store.model.User
import com.firefly.kotlin.ext.db.AsyncTransactionalManager
import com.firefly.kotlin.ext.db.asyncQueryForObject
import com.firefly.utils.StringUtils

/**
 * @author Pengtao Qiu
 */
@Component
class UserDAOImpl : UserDAO {

    @Inject
    private lateinit var db: AsyncTransactionalManager

    override suspend fun getByName(name: String): User {
        if (!StringUtils.hasText(name)) {
            throw IllegalArgumentException("The username is required")
        }

        val sql = "select * from `coffee_store`.`user` where `name` = ?"
        return db.execSQL { it.asyncQueryForObject<User>(sql, name) }
    }

}